src/Internal/Cardano/Types/Transaction.purs

```diff
@@ -788,26 +814,60 @@ derive newtype instance EncodeAeson Vkey
 instance Show Vkey where
   show = genericShow
 
-newtype PublicKey = PublicKey Bech32String
+newtype PublicKey = PublicKey RawBytes
+
+mkPubKey :: Bech32String -> Maybe PublicKey
+mkPubKey = map (PublicKey <<< bytesFromPublicKey) <<< publicKeyFromBech32
+
+mkFromCslPubKey :: Serialization.PublicKey -> PublicKey
+mkFromCslPubKey = PublicKey <<< bytesFromPublicKey
 
-derive instance Generic PublicKey _
-derive instance Newtype PublicKey _
+convertPubKey :: PublicKey -> Serialization.PublicKey
+convertPubKey (PublicKey bs) = unsafePartial $ fromJust <<< fromBytes <<< unwrap
+  $ bs
+
+-- avoid wrap and unwrap being used unsafely
+
+-- derive instance Generic PublicKey _
+-- derive instance Newtype PublicKey _
 derive newtype instance Eq PublicKey
 derive newtype instance Ord PublicKey
 derive newtype instance EncodeAeson PublicKey
 
+instance ToData PublicKey where
+  toData = toData <<< bytesFromPublicKey <<< convertPubKey
+
+{- unsafePartial $ fromJust <<<
+(map toData <<< bytesFromPublicKey <=< publicKeyFromBech32 <<< unwrap) -}
+
+instance FromData PublicKey where
+  fromData = map mkFromCslPubKey <<< fromBytes <=< fromData
+
 instance Show PublicKey where
-  show = genericShow
+  show pk = "PublicKey " <> (bech32FromPublicKey <<< convertPubKey $ pk)
+
+newtype Ed25519Signature = Ed25519Signature RawBytes
 
-newtype Ed25519Signature = Ed25519Signature Bech32String
+mkEd25519Signature :: Bech32String -> Maybe Ed25519Signature
+mkEd25519Signature = map (Ed25519Signature <<< wrap <<< toBytes <<< asOneOf) <<<
+  ed25519SignatureFromBech32
 
-derive instance Generic Ed25519Signature _
+mkFromCslEd25519Signature :: Serialization.Ed25519Signature -> Ed25519Signature
+mkFromCslEd25519Signature = Ed25519Signature <<< wrap <<< toBytes <<< asOneOf
+
+convertEd25519Signature :: Ed25519Signature -> Serialization.Ed25519Signature
```

Comment:
This function does not belong here. Please move it to somewhere in Serialization. In general, `Serialization.*` should depend on cardano types, and not the other way.

Reply:
You mean the convert fn? Its not really straight forward because of circular deps. I'll give it a go.

Reply:
Now I understand the reason we have to do it that way: the Show instance which uses bech32, which is produced by CSL